Update warning about ignoring cached namespaces

stephprince opened the pull request:
## Motivation

Fix #1018.

Combine the warnings about ignoring cached namespaces into a single warning, and only warn for core namespaces if the cached version is newer than the loaded version. 

## How to test the behavior?
with an example file from the pynwb test suite:
```python
from pynwb import NWBHDF5IO

io = NWBHDF5IO("tests/back_compat/2.1.0_nwbfile_with_extension.nwb", "r")
io.read()
```

```
UserWarning: Ignoring the following cached namespace(s) because another version is already loaded:
hdmf-experimental - cached version: 0.2.0, loaded version: 0.5.0
Ignore this warning if these versions are compatible.
```

## Checklist

- [x] Did you update `CHANGELOG.md` with your changes?
- [X] Does the PR clearly describe the problem and the solution?
- [X] Have you reviewed our [Contributing Guide](https://github.com/hdmf-dev/hdmf/blob/dev/docs/CONTRIBUTING.rst)?
- [X] Does the PR use "Fix #XXX" notation to tell GitHub to close the relevant issue numbered XXX when the PR is merged?

stephprince committed:
add util function for version comparison

stephprince committed:
update warning and refactor namespace loading

stephprince committed:
add tests for new namespace warnings

stephprince committed:
remove old warning filtering

stephprince committed:
remove order_deps functions from h5tools

stephprince commented on src/hdmf/spec/namespace.py:
```diff
@@ -508,36 +526,117 @@ def __register_dependent_types_helper(spec, inc_ns, catalog, registered_types):
+        if core_warnings:
+            warning_msg.append(f'{("\n").join(core_warnings)}\nPlease update to the latest package versions.')
+        if other_warnings:
+            warning_msg.append(f'{("\n").join(other_warnings)}\nIgnore this warning if these versions are compatible.')
```
Are there any more actionable steps that we can recommend users to take here if they are not compatible?

rly replied:
For the core warnings, your instructions sound good.

For the other warnings, I would say "The loaded extension(s) may not be compatible with the cached extension(s) in the file. Please check the extension documentation and ignore this warning if these versions are compatible."
What do you think?

In the future, when we have the ability to unload namespaces/extensions, we can update this warning.

stephprince replied:
Those instructions sound good to me! And yes we can provide instructions to unload namespaces when we add those capabilities.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

stephprince commented on src/hdmf/utils.py:
```diff
@@ -876,6 +876,16 @@ def is_ragged(data):
 
     return False
 
+def is_newer_version(version_a, version_b):
+    # this method could be replaced by packaging.version if packaging is added as a dependency
+    for a, b in zip(version_a, version_b):
+        if a > b:
+            return True
+        elif a < b:
+            return False
+
+    return False
```
the packaging library has more robust version comparison functions, but I wasn't sure if we wanted to add it as a dependency just for this case.

rly replied:
I don't think we need to just for this case. We decided against it in pynwb and previously in hdmf as well.

stephprince committed:
update CHANGELOG

stephprince committed:
remove backslashes from f-strings for older python versions

stephprince committed:
fix spelling in comment

stephprince committed:
Merge branch 'dev' into update-namespace-warning

rly commented:
I would really like to not show the hdmf-experimental warning but I know we do not ensure backwards compatibility between schema versions there. What if we also hide the warning if none of the data types in the namespace are used in the file? Most files do not involve data types used in hdmf-experimental.

On rewrite/append, the loaded namespaces are written to the file, so we should not have a problem with writing new data that is incompatible with old namespaces.

stephprince commented:
> I would really like to not show the hdmf-experimental warning but I know we do not ensure backwards compatibility between schema versions there. What if we also hide the warning if none of the data types in the namespace are used in the file? Most files do not involve data types used in hdmf-experimental.

I think that's a good idea! I agree it would be nice to not display that warning unless necessary. 

Are there any places you can point me to in the code that get all the data types used in the file? I am thinking this would be an optional argument to `NamespaceCatalog.load_namespaces` since the NamespaceCatalog seems isolated from the file object?

rly commented:
After talking with @oruebel , we realized that it would be hard or hacky to get all the data types used in the file. The hacky way is to iterate through all groups and datasets in the file and look for the `neurodata_type` or `data_type` attributes, but this involves iterating through the whole file just to check for a warning. This iteration is expensive, especially for streaming data.

Another approach we discussed is to have the IO object build all the builders before loading the namespace from the file. Then we can iterate through the in-memory builders. But this requires refactoring the logic in NWBHDF5IO and HDF5IO, which currently takes in a BuildManager that has the loaded namespace. This would be complicated.

The primary use case for the hdmf-experimental warning is to warn users of incompatible/breaking changes introduced in, say 0.4.0 when the data was written in an older version, say 0.3.0. This really only impacts `ExternalResources`. (`EnumData` is also in hdmf-experimental but has not changed). To my knowledge, no one is using older versions of `ExternalResources` (and perhaps also the latest version), it is relatively stable now, and no one has yet raised an issue about incompatibilities. Therefore the impact of just ignoring this warning is minimal.

So, I would say, let's include hdmf-experimental into the "core" set of namespaces to ignore from hdmf.

stephprince commented:
> So, I would say, let's include hdmf-experimental into the "core" set of namespaces to ignore from hdmf.

Thanks for explaining in detail. That sounds good to me - I'll add the hdmf-experimental namespace to the `NamespaceCatalog.__core_namespace` list. 

Note that the pynwb [`__init__.py`](https://github.com/NeurodataWithoutBorders/pynwb/blob/1b613aa0b1c42be3eebc2c7bab87c157df081512/src/pynwb/__init__.py#L90) file will also need to be updated to add the "core" core namespace and fully eliminate the pynwb/hdmf namespace related warnings when the cached namespace version is not newer. (In the example output I demonstrated above with "tests/back_compat/2.1.0_nwbfile_with_extension.nwb", I think I was accidentally using a cached typemap in which I had added the "core" namespace in pynwb for testing purposes). 

Edit: see https://github.com/NeurodataWithoutBorders/pynwb/issues/2064 as the open issue to add "core" as a core namespace

stephprince committed:
add hdmf-experimental to core_namespace list

stephprince commented:
@rly this is ready for you to look at again

rly committed:
Merge branch 'dev' into update-namespace-warning

rly commented on src/hdmf/spec/namespace.py:
```diff
@@ -229,13 +229,23 @@ class NamespaceCatalog:
+        core_namespace = getargs('core_namespace', kwargs)
+        if core_namespace is None:
+            from ..common import CORE_NAMESPACE, EXP_NAMESPACE
+            self.__core_namespace = [CORE_NAMESPACE, EXP_NAMESPACE]
```
```suggestion
            # NOTE: even though HDMF does not guarantee backwards compatibility with schema
            # using an older version of the experimental namespace, in practice, this has not been
            # an issue, and it is costly to determine whether there is an incompatibility before issuing
            # a warning. so, we ignore the experimental namespace warning by default. 
            # see https://github.com/hdmf-dev/hdmf/pull/1258
            self.__core_namespace = [CORE_NAMESPACE, EXP_NAMESPACE]
```

rly committed:
Update src/hdmf/spec/namespace.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

rly commented on src/hdmf/utils.py:
```diff
@@ -876,6 +876,16 @@ def is_ragged(data):
 
     return False
 
+def is_newer_version(version_a, version_b):
```
It looks like the inputs in the code are strings, and then this code splits the strings into lists of characters and compares them element by element, so:
```
>>> is_newer_version("2.0.0", "10.0.0")
True
```
which is incorrect. You can use part of this code to get the version string as a list of ints: https://github.com/NeurodataWithoutBorders/pynwb/blob/1b613aa0b1c42be3eebc2c7bab87c157df081512/src/pynwb/io/utils.py#L7
and then this code should run fine.

stephprince replied:
thanks for catching that!

rly commented on src/hdmf/spec/namespace.py:
```diff
@@ -229,13 +229,28 @@ class NamespaceCatalog:
+
+        core_namespace = getargs('core_namespace', kwargs)
+        if core_namespace is None:
+            from ..common import CORE_NAMESPACE, EXP_NAMESPACE
+            # NOTE: even though HDMF does not guarantee backwards compatibility with schema
+            # using an older version of the experimental namespace, in practice, this has not been
+            # an issue, and it is costly to determine whether there is an incompatibility before issuing
+            # a warning. so, we ignore the experimental namespace warning by default.
+            # see https://github.com/hdmf-dev/hdmf/pull/1258
+            self.__core_namespace = [CORE_NAMESPACE, EXP_NAMESPACE]
```
I think this should be handled in https://github.com/hdmf-dev/hdmf/blob/dev/src/hdmf/common/__init__.py#L236

This class should not know anything about hdmf.common.

To make this work, I think the `core_namespace` string argument should change to `core_namespaces` as a list of strings.

rly commented on tests/unit/spec_tests/test_load_namespace.py:
```diff
@@ -278,11 +286,55 @@ def test_catch_dup_name_same_version(self):
+        ns_catalog.merge(self.ns_catalog)
+
+        # test no warning if loading older namespace than one already loaded
+        ns_catalog.load_namespaces(os.path.join(self.tempdir, self.ns_path1))
```
Although we try to look at all warnings returned by tests, in CI, we do not always. I think it would be better to follow the pattern from one of the tests above to test that the namespace warning does not appear.

stephprince replied:
That makes sense, I wasn't sure the best approach on this one. It also felt risky that when I changed the warning message, the test checking for the exact warning string would continue to pass even when the case was raising a warning... but I guess either approach has the potential to miss some things.

rly replied:
Good point. I suggest we re-raise any caught warnings in the for loop `for w in ws`. Could you update that in all places we use that pattern?

stephprince committed:
update version comparison function

stephprince committed:
move core_namespace list definition to hdmf common init

stephprince committed:
add warning comparison to tests

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

stephprince commented:
@rly I think I addressed all of your comments

I believe the test failures are due to an ImportError issue with the schemasheets package. The warning in the logs is a little confusing because schemasheets is actually installed, but does not seem to be compatible with the latest versions of `linkml_runtime>=1.9.0`. I'll open up an issue on that repo (edit: see https://github.com/linkml/schemasheets/issues/147).

rly commented:
Thanks! I made a couple minor comments. Otherwise looks good.

stephprince committed:
update warning message for extension namespace compatibility

stephprince committed:
Merge branch 'dev' into update-namespace-warning

stephprince commented:
@rly I think I addressed all your remaining comments! This should be ready to merge

rly commented:
Looks good. Thank you! Squash and merge anytime.

rly commented:
I'll just do it since Matt and I are managing this package and the release.

h-mayorquin commented:
This is great! Thanks for moving this forward!
